cmd/oci-image-tool/validate.go

```diff
@@ -0,0 +1,207 @@
+	typeImage,
+	typeManifest,
+	typeManifestList,
+	typeConfig,
+}
+
+type validateCmd struct {
+	stdout *log.Logger
+	stderr *log.Logger
+	typ    string // the type to validate, can be empty string
+}
+
+func newValidateCmd(stdout, stderr *log.Logger) *cobra.Command {
+	v := &validateCmd{
+		stdout: stdout,
+		stderr: stderr,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "validate FILE...",
+		Short: "Validate one or more image files",
+		Run:   v.Run,
+	}
+
+	cmd.Flags().StringVar(
+		&v.typ, "type", "",
+		fmt.Sprintf(
+			`Type of the file to validate. If unset, oci-image-tool will try to auto-detect the type. One of "%s"`,
+			strings.Join(validateTypes, ","),
+		),
+	)
+
+	return cmd
+}
+
+func (v *validateCmd) Run(cmd *cobra.Command, args []string) {
+	if len(args) < 1 {
+		v.stderr.Printf("no files specified")
+		if err := cmd.Usage(); err != nil {
+			v.stderr.Println(err)
+		}
+		os.Exit(1)
+	}
+
+	var exitcode int
+	for _, arg := range args {
+		err := v.validatePath(arg)
+
+		if err == nil {
+			v.stdout.Printf("file %s: OK", arg)
+			continue
+		}
+
+		var errs []error
+		if verr, ok := errors.Cause(err).(oci.ValidationError); ok {
+			errs = verr.Errs
+		} else {
+			v.stderr.Printf("file %s: validation failed: %v", arg, err)
+			exitcode = 1
+			continue
+		}
+
+		for _, err := range errs {
+			v.stderr.Printf("file %s: validation failed: %v", arg, err)
+		}
+
+		exitcode = 1
+	}
+
+	os.Exit(exitcode)
+}
+
+func (v *validateCmd) validatePath(name string) error {
+	var err error
+	typ := v.typ
+
+	if typ == "" {
+		if typ, err = autodetect(name); err != nil {
+			return errors.Wrap(err, "unable to determine type")
+		}
+	}
+
+	switch typ {
+	case typeManifest:
+		if err := withFile(name, oci.ValidateManifest); err != nil {
+			return err
+		}
+
+		return nil
+	case typeManifestList:
+		if err := withFile(name, oci.ValidateManifestList); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("type %q unimplemented", typ)
+}
+
+// withFile opens the named file for reading and executes the given function fn if no error occured.
+// If the given function returns an error it will be returned.
+func withFile(name string, fn func(f io.Reader) error) error {
+	f, err := os.Open(name)
+	if err != nil {
+		return errors.Wrap(err, "unable to open file")
+	}
+	defer f.Close()
+
+	return fn(f)
+}
+
+// autodetect returns one of the type.. constants of the content of the given path
+// or an error if the type could not be resolved.
+func autodetect(path string) (string, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to access path") // err from os.Stat includes path name
+	}
+
+	if fi.IsDir() {
+		return typeImageLayout, nil
+	}
+
+	f, err := os.Open(path)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to open file") // os.Open includes the filename
+	}
+	defer f.Close()
+
+	buf, err := ioutil.ReadAll(io.LimitReader(f, 512)) // read up to 512 bytes for content detection
```

Comment:
I would just say "read some initial bytes to detect content" instead of putting the amount in the comment